feat: Single discord channel import UI flow

caybro opened the pull request:
Closes #12039

### What does the PR do

Implements the UI changes in CreateChannelPopup needed to be able to import a single Discord channel

### Affected areas

CreateChannelPopup

### Screenshot of functionality (including design for comparison)

- [x] I've checked the design and this PR matches it

[Záznam obrazovky z 2023-09-04 11-23-13.webm](https://github.com/status-im/status-desktop/assets/5377645/9c5ed0b2-34e5-4aa8-88d0-011f11b8968e)

status-im-auto commented:
### Jenkins Builds
| :grey_question: | Commit | :hash: | Finished (UTC) | Duration | Platform | Result |
|-|-|-|-|-|-|-|
| :heavy_check_mark: | d80ec4f9 | [#1](https://ci.status.im/job/status-desktop/job/branches/job/linux/job/x86_64/job/tests-imports/job/PR-12060/1/) | 2023-09-01 08:04:59 | ~4 min | `tests/imports` | [:page_facing_up:`log`](https://ci.status.im/job/status-desktop/job/branches/job/linux/job/x86_64/job/tests-imports/job/PR-12060/1//consoleText) |

caybro commented:
@borismelnik I started the UI work here already, you can base yours on top of it. I added some stubs CommunitiesStore.qml and some TODOs in the NIM files, feel free to take it on from there. The UI work is not completely finished yet but it's already somehow usable

caybro commented:
@borismelnik also now added some NIM methods to handle the UI selection

caybro commented:
@jrainville @borismelnik the UI is now usable up to the last step which should be showing the progress of the actual import (to be implemented in NIM/status-go of course ;)

CommunitiesStore.qml has 2 stub methods to be filled with actual backend calls: `requestImportDiscordChannel(args,from)` and `requestCancelDiscordChannelImport(id)` 

All of the other logic could be imho left the same, as if we were importing the whole community:
```qml
    property var discordFileList: root.communitiesModuleInst.discordFileList
    property var discordCategoriesModel: root.communitiesModuleInst.discordCategories
    property var discordChannelsModel: root.communitiesModuleInst.discordChannels
    property int discordOldestMessageTimestamp: root.communitiesModuleInst.discordOldestMessageTimestamp
    property bool discordDataExtractionInProgress: root.communitiesModuleInst.discordDataExtractionInProgress
    property int discordImportErrorsCount: root.communitiesModuleInst.discordImportErrorsCount
    property int discordImportWarningsCount: root.communitiesModuleInst.discordImportWarningsCount
    property int discordImportProgress: root.communitiesModuleInst.discordImportProgress
    property bool discordImportInProgress: root.communitiesModuleInst.discordImportInProgress
    property bool discordImportCancelled: root.communitiesModuleInst.discordImportCancelled
    property bool discordImportProgressStopped: root.communitiesModuleInst.discordImportProgressStopped
    property int discordImportProgressTotalChunksCount: root.communitiesModuleInst.discordImportProgressTotalChunksCount
    property int discordImportProgressCurrentChunk: root.communitiesModuleInst.discordImportProgressCurrentChunk
    property string discordImportCommunityId: root.communitiesModuleInst.discordImportCommunityId
    property string discordImportCommunityName: root.communitiesModuleInst.discordImportCommunityName
    property string discordImportChannelName: root.communitiesModuleInst.discordImportChannelName // TODO implement in NIM
    property url discordImportCommunityImage: root.communitiesModuleInst.discordImportCommunityImage
    property bool discordImportHasCommunityImage: root.communitiesModuleInst.discordImportHasCommunityImage
    property var discordImportTasks: root.communitiesModuleInst.discordImportTasks
```

The progress dialog/banner relies on these ^^

borismelnik commented on src/app/modules/main/communities/view.nim:
```diff
@@ -609,6 +613,30 @@ QtObject:
+  proc discordImportChannelChanged*(self: View) {.signal.}
+
+  proc toggleOneDiscordChannel*(self: View, id: string) {.slot.} =
+    let item = self.discordChannelsModel.getItem(id)
```
`getItem` can return `null` maybe good to check it

micieslak commented on ui/app/AppLayouts/Communities/popups/CreateChannelPopup.qml:
```diff
@@ -70,293 +178,456 @@ StatusDialog {
+                }
+                IssuePill {
+                    type: root.communitiesStore.discordImportErrorsCount ? IssuePill.Type.Error : IssuePill.Type.Warning
+                    count: {
+                      if (root.communitiesStore.discordImportErrorsCount > 0) {
+                        return root.communitiesStore.discordImportErrorsCount
+                      }
+                      if (root.communitiesStore.discordImportWarningsCount > 0) {
+                        return root.communitiesStore.discordImportWarningsCount
+                      }
+                      return 0
                     }
```
```suggestion
                    count: root.communitiesStore.discordImportErrorsCount
                        ? root.communitiesStore.discordImportErrorsCount
                        : root.communitiesStore.discordImportWarningsCount
```
or the shortest version:
```suggestion
                    count: count: root.communitiesStore.discordImportErrorsCount || root.communitiesStore.discordImportWarningsCount
```

micieslak commented on ui/app/AppLayouts/Communities/popups/CreateChannelPopup.qml:
```diff
@@ -70,293 +178,456 @@ StatusDialog {
+            spacing: 24
+
+            readonly property bool canGoNext: root.communitiesStore.discordChannelsModel.hasSelectedItems
+            readonly property var nextAction: function () {
```
couldn't it be just:
```suggestion
            function nextAction() {
```
?

caybro replied:
Nope, it's used by the handlers below...

micieslak replied:
Hmm, does it change anything? Function can be accessed by name in the same as a regular property, right?

caybro replied:
Hmm, probably it doesn't change anything, it's accessed as:
```qml
const nextAction = currentItem.nextAction
if (typeof(nextAction) == "function") {
    return nextAction()
}
```

micieslak commented on ui/app/AppLayouts/Communities/popups/CreateChannelPopup.qml:
```diff
@@ -70,293 +178,456 @@ StatusDialog {
+            padding: 0
+
+            function scrollBackUp() {
+                vScrollBar.setPosition(0)
```
shouldn't it be just
```suggestion
                vScrollBar.position = 0
```
?
I can't find setPosition in a documented api

caybro replied:
Not my code, dunno

micieslak replied:
yeah, but maybe good time to fix it once spotted

caybro replied:
Probably some copy&pasted code (not mine :), but it definitely works

micieslak commented on ui/app/AppLayouts/Communities/popups/CreateChannelPopup.qml:
```diff
@@ -70,293 +178,456 @@ StatusDialog {
-            width: scrollView.availableWidth
-            spacing: 0
+            id: fileListView
+            objectName: "discordFileListView" // !!! DON'T CHANGE, clearFilesButton depends on this
```
what about sth like that:
```suggestion
            readonly property bool isFileListView: true
```
?
Then there is no risk of interfering with tests relying on objectName

micieslak commented on ui/app/AppLayouts/Communities/popups/CreateChannelPopup.qml:
```diff
@@ -70,293 +178,456 @@ StatusDialog {
-                    if (root.channelEmoji === "") {
-                        input.letterIconName = text;
+                spacing: 12
+                StatusBaseText {
```
Texts like that should be prepared for input longer then initial English text. Even with slightly longer text positioning is broken here.

micieslak replied:
Just replace "Select Discord channel JSON files to import" text with a bit longer text. The layout is broken then. It may easily happen for translations where given sentence is longer than in English.

caybro replied:
Ah I see! Will add an elide/wrap there

caybro replied:
Fixed @micieslak, thanks!

micieslak commented on storybook/pages/CreateChannelPopupPage.qml:
```diff
@@ -19,15 +19,105 @@ SplitView {
+                }
+
+                function removeFileListItem(path) {
+                    for (var i = 0; i < discordFileList.count; i++) {
```
`var` -> `let`, basically in most cases there is no reason to use `var` in modern js. block-scoped `let`/`const` should be used instead

micieslak commented on storybook/pages/CreateChannelPopupPage.qml:
```diff
@@ -19,15 +19,105 @@ SplitView {
+
+                function removeFileListItem(path) {
+                    for (var i = 0; i < discordFileList.count; i++) {
+                        var item = discordFileList.get(i)
```
const

micieslak commented on ui/app/AppLayouts/Communities/popups/CreateChannelPopup.qml:
```diff
@@ -43,24 +46,129 @@ StatusDialog {
+        enabled: typeof(currentItem.canGoNext) == "undefined" || currentItem.canGoNext
+        loading: root.communitiesStore.discordDataExtractionInProgress
+        onClicked: {
+            let nextAction = currentItem.nextAction
```
```suggestion
            const nextAction = currentItem.nextAction
```

micieslak commented on ui/app/AppLayouts/Communities/popups/CreateChannelPopup.qml:
```diff
@@ -43,24 +46,129 @@ StatusDialog {
+        }
+    }
+
+    readonly property var clearFilesButton: StatusButton {
```
```suggestion
    readonly property StatusButton clearFilesButton: StatusButton {
```
or maybe no property at all, just regular object referred by id? Is it necessary to expose it?

caybro replied:
It's needed for the footer buttons (`rightButtons` property)

micieslak reviewed (commented):
Nice job in general, just some comments and questions for further consideration.

General:
- it's impossible to add files from various folders, bc the consecutive selections replace the whole list instead of appending
- In the design sizing and scrolling behavior is a bit different:

Design - popup filling the window height - margins, gray separators visible when scrolling
![Screenshot from 2023-09-05 11-37-47](https://github.com/status-im/status-desktop/assets/20650004/e4578e7e-a707-47bc-a119-50f66408ba96)

Implementation:
![Screenshot from 2023-09-05 11-36-41](https://github.com/status-im/status-desktop/assets/20650004/8b82a28b-a9b7-402c-b2de-3596b140ba9e)

Another detail is text in the `Validate` button - one number when all files are ok.

- The approach for invoking popup from context menu is a bit inconsistent with "Add channels" which gives possibility to create only regular empty channel. It's easy to miss possibility of importing. Probably it's a question to @benjthayer 
![Screenshot from 2023-09-05 11-43-22](https://github.com/status-im/status-desktop/assets/20650004/3c48d973-21a8-4c0f-92f9-c7463cd57ce5)

caybro commented:
> Nice job in general, just some comments and questions for further consideration.
> 
> General:
> 
> * it's impossible to add files from various folders, bc the consecutive selections replace the whole list instead of appending

Yeah... it's never been designed with this in mind (also the NIM backend is not prepared for this); perhaps worth filing a separate issue?

> * In the design sizing and scrolling behavior is a bit different:

Yup, will have a look

> Another detail is text in the `Validate` button - one number when all files are ok.

Alright, one more tiny difference from the previous design(s)

caybro commented on src/app/modules/main/communities/models/discord_categories_model.nim:
```diff
@@ -107,3 +107,10 @@ QtObject:
       self.items[idx].selected = true
       self.dataChanged(index, index, @[ModelRole.Selected.int])
+
+  proc selectOneItem*(self: DiscordCategoriesModel, id: string) =
```
This is impossible in NIM imo :) Also, here we have a unique selection, one item at most (radio buttons)

micieslak replied:
Imo it is, because than it lives exclusively on UI side. It also gives full independence between the actual model and the selection that is done on top. The qml type is https://doc.qt.io/qt-5/qml-qtqml-models-itemselectionmodel.html, it nicely manages selection, including observing the underlying object and reacting on removals, moves and so on. Additional profit is that UI side is really on ui side and when used in Storybook, you don't have to reimplement selection to have the component functional.

caybro replied:
Ok then, if you insist - please file a separate issue for the backend part of the corresponding epic; I merely extended the existing NIM code to make the UI functional

micieslak replied:
No, I'm not insisting, just pointing that it's not impossible. Moving it to a separate task is definitely a good idea.

micieslak replied:
here you go: https://github.com/status-im/status-desktop/issues/12099 :)

caybro commented on storybook/pages/CreateChannelPopupPage.qml:
```diff
@@ -19,15 +19,105 @@ SplitView {
+            }
         }
 
         CreateChannelPopup {
```
Yeah, I initially planned to have it fully working in storybook too/first but then realized it would basically mean reimplementing the whole logic and models from NIM here in JS :/

caybro commented:
> > * In the design sizing and scrolling behavior is a bit different:
> 
> Yup, will have a look
> 

Fixed the scrolling, couldn't fix the overall popup sizing - it's an issues with the StatusStackModal, it simply gets the `implicitHeight` from the initial page and ignores the rest :/

> > Another detail is text in the `Validate` button - one number when all files are ok.
> 

Fixed

Pls have a second look @micieslak :)

alexjba commented on src/app/modules/main/communities/models/discord_categories_model.nim:
We need to translate these messages into user friendly errors:

Incorrect format:
<img width="1840" alt="Screenshot 2023-09-06 at 10 53 34" src="https://github.com/status-im/status-desktop/assets/47811206/b3ec6cad-1b50-4443-9bf5-44a191a9123f">

Missing file:
<img width="1840" alt="Screenshot 2023-09-06 at 10 56 06" src="https://github.com/status-im/status-desktop/assets/47811206/82a4edcd-fcb6-47af-84ef-a8d70aabe01c">

Empty file:
<img width="1840" alt="Screenshot 2023-09-06 at 11 06 25" src="https://github.com/status-im/status-desktop/assets/47811206/ca932c24-7101-444f-8204-bafd0d70ce3e">

caybro replied:
These are coming from backend, impossible to do :/

alexjba replied:
It means we're using the same approach as the login flow. IMO this is not a good approach. We should receive error codes and display the proper error in the UI based on error code. This error message is probably the exception description. It can't be translated.

caybro replied:
Please file a separate issue for the backend tasks in respective epic: https://github.com/status-im/status-desktop/issues/12018

alexjba replied:
Done https://github.com/status-im/status-desktop/issues/12098

alexjba commented on src/app/modules/main/communities/models/discord_categories_model.nim:
I see in the [figma design](https://www.figma.com/file/qHfFm7C9LwtXpfdbxssCK3/Kuba⎜Desktop---Communities?type=design&node-id=37454-338477&mode=design&t=JJyT0AuuyT9upTIH-0) we should support drag and drop. Currently it's not working. Looks like the statusChatInput steals the drop area.

<img width="1840" alt="Screenshot 2023-09-06 at 11 07 37" src="https://github.com/status-im/status-desktop/assets/47811206/3f8389fc-e0fa-41ea-a9b8-29cb65e7c26e">

alexjba commented on src/app/modules/main/communities/models/discord_categories_model.nim:
The layout gets broken by the following flow:

1. Add one invalid json file
2. Validate the file -> Showing the error
3. Hit the `x` next to the file name to remove it
4. Broken layout 

<img width="1840" alt="Screenshot 2023-09-06 at 10 56 40" src="https://github.com/status-im/status-desktop/assets/47811206/3ec7e0a2-350c-419f-aab9-5608eabe6da7">

caybro replied:
Ah, will fix!

caybro replied:
Fixed:

![image](https://github.com/status-im/status-desktop/assets/5377645/3ec38338-3a23-4642-847e-d056d3e9d0b2)

alexjba commented on src/app/modules/main/communities/models/discord_categories_model.nim:
The wiki URL seems to be outdated:
https://github.com/Tyrrrz/DiscordChatExporter/wiki

alexjba commented on src/app/modules/main/communities/models/discord_categories_model.nim:
I see the `DiscordChatExporter` URL points to the main GitHub page. This is great as long as it's guaranteed that there's not going to be breaking changes in  `DiscordChatExporter` json formatting. My suggestion would be to use the URL pointing to the release we know for sure it's working. This way the user will always get a `DiscordChatExporter` version that is compatible to our current implementation.

WDYT?

caybro replied:
No idea there... @0x-r4bbit ?

0x-r4bbit replied:
Sure go ahead. Fixed release sounds reasonable to me

caybro replied:
No problem with that, they just make new releases pretty often and it's hard to say which one is a "good" one

caybro replied:
OK, going with the latest "stable" (https://github.com/Tyrrrz/DiscordChatExporter/releases/tag/2.40.4)

alexjba commented on src/app/modules/main/communities/models/discord_categories_model.nim:
Cannot add files from multiple folders because on each selection the old files are removed from the list.
In order for me to select files from multiple folders, I need to open the file explorer multiple times and select the files from different folders.


https://github.com/status-im/status-desktop/assets/47811206/1f18aa4a-150d-41f5-aea9-71c5674c6dfb

caybro replied:
Yup, already pointed out by @micieslak, the backend simply wasn't designed with this in mind

alexjba commented on src/app/modules/main/communities/models/discord_categories_model.nim:
Is there a different task for the `permissions` section?

<img width="421" alt="Screenshot 2023-09-06 at 11 29 38" src="https://github.com/status-im/status-desktop/assets/47811206/7b4a7a55-dfdf-4bb6-9463-96f1683e178f">

caybro replied:
I'm not aware of any, maybe @benjthayer has an idea? In any case, this "permissions" feature isn't part of this task imo

alexjba reviewed (commented):
Some questions/findings before going into the code changes:

alexjba commented on ui/app/AppLayouts/Communities/popups/DiscordImportProgressContents.qml:
```diff
@@ -329,8 +333,9 @@ StatusScrollView {
             text: d.status === DiscordImportProgressContents.ImportStatus.InProgress ?
-                      qsTr("This process can take a while. Feel free to hide this window and use Status normally in the meantime. We’ll notify you when the Community is ready for you.") :
+                      qsTr("This process can take a while. Feel free to hide this window and use Status normally in the meantime. We’ll notify you when the %1 is ready for you.") :
                       qsTr("If there were any issues with your import you can upload new JSON files via the community page at any time.")
```
The argument is missing

caybro replied:
Fixed

alexjba reviewed (commented):
Nice work! 

Minor findings below

alexjba commented:
One more thing. I think the color is wrong in dark mode on some controls.

<img width="1840" alt="Screenshot 2023-09-06 at 12 31 35" src="https://github.com/status-im/status-desktop/assets/47811206/61131c18-dfd5-4bf3-be0b-b00991fc31f8">
<img width="1840" alt="Screenshot 2023-09-06 at 12 31 03" src="https://github.com/status-im/status-desktop/assets/47811206/a358f4bd-21b2-473e-903c-0a69974eda21">

caybro commented on ui/app/AppLayouts/Communities/popups/CreateChannelPopup.qml:
```diff
@@ -14,15 +14,18 @@ import StatusQ.Controls 0.1
     id: root
-    width: 480
-    height: 509
 
```
Why? :)

alexjba replied:
Usually a popup with `X` button doesn't use auto-close when clicking outside.

caybro replied:
All our popups behave that way, even if there's an "X" button

alexjba replied:
Ohh, you're right!

caybro commented:
> One more thing. I think the color is wrong in dark mode on some controls.
> 
> <img alt="Screenshot 2023-09-06 at 12 31 35" width="1840" src="https://user-images.githubusercontent.com/47811206/265964096-61131c18-dfd5-4bf3-be0b-b00991fc31f8.png"> <img alt="Screenshot 2023-09-06 at 12 31 03" width="1840" src="https://user-images.githubusercontent.com/47811206/265964118-a358f4bd-21b2-473e-903c-0a69974eda21.png">

Hmm not sure there... definitely haven't changed anything there; it's the primary/dark variant

alexjba commented:
> > One more thing. I think the color is wrong in dark mode on some controls.
> > <img alt="Screenshot 2023-09-06 at 12 31 35" width="1840" src="https://user-images.githubusercontent.com/47811206/265964096-61131c18-dfd5-4bf3-be0b-b00991fc31f8.png"> <img alt="Screenshot 2023-09-06 at 12 31 03" width="1840" src="https://user-images.githubusercontent.com/47811206/265964118-a358f4bd-21b2-473e-903c-0a69974eda21.png">
> 
> Hmm not sure there... definitely haven't changed anything there; it's the primary/dark variant

Could be some broken bindings? I think the `Browse files` button should look like `Validate` button. I'm not sure what colours the radio buttons should have, but doesn't look like it's using the dark theme colours.

caybro commented:
@alexjba pls have a second look

caybro commented:
> Could be some broken bindings? I think the `Browse files` button should look like `Validate` button. I'm not sure what colours the radio buttons should have, but doesn't look like it's using the dark theme colours.

Nope, not the same thing; the `Browse` button has this in addition: `type: StatusBaseButton.Type.Primary`

caybro commented:
> > Could be some broken bindings? I think the `Browse files` button should look like `Validate` button. I'm not sure what colours the radio buttons should have, but doesn't look like it's using the dark theme colours.
> 
> Nope, not the same thing; the `Browse` button has this in addition: `type: StatusBaseButton.Type.Primary`

@alexjba compare the first one (Primary) vs the second one (Regular):
![image](https://github.com/status-im/status-desktop/assets/5377645/92da4519-6f8c-4fe6-b7b3-e02ad1f8caad)

caybro commented:
> > > Could be some broken bindings? I think the `Browse files` button should look like `Validate` button. I'm not sure what colours the radio buttons should have, but doesn't look like it's using the dark theme colours.
> > 
> > 
> > Nope, not the same thing; the `Browse` button has this in addition: `type: StatusBaseButton.Type.Primary`
> 
> @alexjba compare the first one (Primary) vs the second one (Regular): ![image](https://user-images.githubusercontent.com/5377645/266303507-92da4519-6f8c-4fe6-b7b3-e02ad1f8caad.png)

OK, the white text on the primary doesn't look correct :) But the blue color is fine

alexjba commented:
> > > > Could be some broken bindings? I think the `Browse files` button should look like `Validate` button. I'm not sure what colours the radio buttons should have, but doesn't look like it's using the dark theme colours.
> > > 
> > > 
> > > Nope, not the same thing; the `Browse` button has this in addition: `type: StatusBaseButton.Type.Primary`
> > 
> > 
> > @alexjba compare the first one (Primary) vs the second one (Regular): ![image](https://user-images.githubusercontent.com/5377645/266303507-92da4519-6f8c-4fe6-b7b3-e02ad1f8caad.png)
> 
> OK, the white text on the primary doesn't look correct :) But the blue color is fine

True. The color on the radio buttons seems correct (at least consistent with other places)

caybro commented:
> > > > > > Could be some broken bindings? I think the `Browse files` button should look like `Validate` button. I'm not sure what colours the radio buttons should have, but doesn't look like it's using the dark theme colours.
> > > > > 
> > > > > 
> > > > > Nope, not the same thing; the `Browse` button has this in addition: `type: StatusBaseButton.Type.Primary`
> > > > 
> > > > 
> > > > @alexjba compare the first one (Primary) vs the second one (Regular): ![image](https://user-images.githubusercontent.com/5377645/266303507-92da4519-6f8c-4fe6-b7b3-e02ad1f8caad.png)
> > > 
> > > 
> > > OK, the white text on the primary doesn't look correct :) But the blue color is fine
> > 
> > 
> > True. The color on the radio buttons seems correct (at least consistent with other places)
> 
> Yup, we need to fix the primary/dark button text color

@alexjba pushed a (separate) fix for that:
![image](https://github.com/status-im/status-desktop/assets/5377645/b2ce980a-4257-4d03-bbbe-95847acf3ad9)

alexjba reviewed (approved):
LGTM! Great work 👍

micieslak reviewed (approved):
Nice!
